crates/wasi/src/filesystem.rs

```diff
@@ -236,45 +252,175 @@ impl Dir {
+        self.wait_ready().await;
+
+        match &mut self.state {
+            ReadState::Idle => {
+                if size == 0 {
+                    return Ok(Bytes::new());
+                }
+
+                let p = self.position;
+                match self
+                    .file
+                    .run_blocking(move |f| Self::blocking_read(f, p, size))
+                    .await
+                {
+                    Ok(b) if b.len() == 0 => {
+                        self.state = ReadState::Closed;
+                        Err(StreamError::Closed)
+                    }
+                    Ok(b) => {
+                        self.position += b.len() as u64;
+                        Ok(b)
+                    }
+                    Err(e) => {
+                        self.state = ReadState::Closed;
+                        Err(StreamError::LastOperationFailed(e.into()))
+                    }
+                }
+            }
+            ReadState::DataAvailable(_) => Ok(self.read_available_data(size)),
+            ReadState::Waiting(_) => Ok(Bytes::new()),
+            ReadState::Error(_) => match mem::replace(&mut self.state, ReadState::Closed) {
+                ReadState::Error(e) => Err(StreamError::LastOperationFailed(e.into())),
+                _ => unreachable!(),
+            },
+            ReadState::Closed => Err(StreamError::Closed),
+        }
+    }
+    async fn cancel(&mut self) {
+        match mem::replace(&mut self.state, ReadState::Closed) {
+            ReadState::Waiting(task) => {
+                // The task was created using `spawn_blocking`, so unless we're
+                // lucky enough that the task hasn't started yet, the abort
+                // signal won't have any effect and we're forced to wait for it
+                // to run to completion.
+                // From the guest's point of view, `input-stream::drop` then
+                // appears to block. Certainly less than ideal, but arguably still
+                // better than letting the guest rack up an unbounded number of
+                // background tasks. Also, the guest is only blocked if
+                // the stream was dropped mid-read, which we don't expect to
+                // occur frequently.
+                task.abort_wait().await;
+            }
+            _ => {}
+        }
     }
 }
+#[async_trait::async_trait]
+impl Subscribe for FileInputStream {
+    async fn ready(&mut self) {
+        if let ReadState::Idle = self.state {
+            // The guest hasn't initiated any read, but is nonetheless waiting
+            // for data to be available. We'll start a read for them:
+
+            const DEFAULT_READ_SIZE: usize = 4096;
+            let p = self.position;
+            self.state = ReadState::Waiting(
+                self.file
+                    .spawn_blocking(move |f| Self::blocking_read(f, p, DEFAULT_READ_SIZE)),
+            );
```

Comment:
I'm a little confused by this, is it required? I would expect that reads are only initiated from the `read` method above (or blocking-read) and the `ready` bits here only wait for completion of an already pending read.

Reply:
> is it required?

Not sure, but the current wasmtime-wasi implementation would push me towards a hesitant "Yes?"
I kinda stole the idea from the stdin implementation that also initiates a new read if there's no data available:

https://github.com/bytecodealliance/wasmtime/blob/c7756bd2654fe1ac795f9dd9d13de41a31cafbac/crates/wasi/src/stdio/worker_thread_stdin.rs#L156-L175

This, combined with the fact that there exist multiple places in the code that blindly assume that calling `.ready().await` will yield readable data:

https://github.com/bytecodealliance/wasmtime/blob/c7756bd2654fe1ac795f9dd9d13de41a31cafbac/crates/wasi/src/host/io.rs#L178-L182
https://github.com/bytecodealliance/wasmtime/blob/c7756bd2654fe1ac795f9dd9d13de41a31cafbac/crates/wasi/src/host/io.rs#L211-L214
https://github.com/bytecodealliance/wasmtime/blob/c7756bd2654fe1ac795f9dd9d13de41a31cafbac/crates/wasi/src/host/io.rs#L231-L234
https://github.com/bytecodealliance/wasmtime/blob/c7756bd2654fe1ac795f9dd9d13de41a31cafbac/crates/wasi/src/stream.rs#L153-L154

Reply:
Hm ok, I'm not sure if this will survive the test of time but following suit with stdin doesn't seem unreasonable. This shouldn't get hit most of the time for the time being as WASIp1 shouldn't use this. Let's go ahead and land it 👍